apps/vault.go

```diff
@@ -1,10 +1,299 @@
+		r.MetricsPath = defaultVaultMetricsPath
+	}
+
+	if r.MetricsPath == "" {
+		r.Scheme = defaultVaultScheme
+	}
+
+	tlsConfig := r.TLSConfig(true)
+
+	scrapeConfig := map[string]interface{}{
+		"job_name":        "vault",
+		"scrape_interval": r.CollectionIntervalString(),
+		"metrics_path":    r.MetricsPath,
+		"static_configs": []map[string]interface{}{{
+			"targets": []string{r.Endpoint},
+		}},
+	}
+
+	if r.Token != "" {
+		scrapeConfig["authorization"] = map[string]interface{}{
+			"credentials": r.Token,
+			"type":        "Bearer",
+		}
+	}
+	if tlsConfig["insecure"] == false {
+		scrapeConfig["scheme"] = "https"
+	}
+	delete(tlsConfig, "insecure")
+	scrapeConfig["tls_config"] = tlsConfig
+
+	includeMetrics := []string{}
+	queries := []otel.TransformQuery{}
+
+	storageMetricTransforms, newStorageMetricNames := r.addStorageMetrics()
+	metricRenewRevokeTransforms, newRenewRevokeNames := r.getRenewRevokeMetricsTransforms()
+	metricDetailTransforms, newMetricNames := r.getMetricTransforms()
+
+	includeMetrics = append(includeMetrics, newStorageMetricNames...)
+	includeMetrics = append(includeMetrics, newMetricNames...)
+	includeMetrics = append(includeMetrics, newRenewRevokeNames...)
+
+	queries = append(queries, storageMetricTransforms...)
+	queries = append(queries, metricRenewRevokeTransforms...)
+	queries = append(queries, metricDetailTransforms...)
+
+	return []otel.Pipeline{{
+		Receiver: otel.Component{
+			Type: "prometheus",
+			Config: map[string]interface{}{
+				"config": map[string]interface{}{
+					"scrape_configs": []map[string]interface{}{
+						scrapeConfig,
+					},
+				},
+			},
+		},
+		Processors: []otel.Component{
+			otel.TransformationMetrics(queries...),
+			otel.MetricsFilter(
+				"include",
+				"strict",
+				includeMetrics...,
+			),
+			// This is currently needed along side the newer transform processor as the new processor doesn't currently support toggling scalar data types.
+			// Issue: https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/11810
+			otel.MetricsTransform(
+				otel.UpdateMetric(
+					"vault.audit.response.failed",
+					otel.ToggleScalarDataType,
+				),
+				otel.UpdateMetric(
+					"vault.audit.request.failed",
+					otel.ToggleScalarDataType,
+				),
+				otel.UpdateMetric(
+					"vault.token.lease.count",
+					otel.ToggleScalarDataType,
+				),
+				otel.UpdateMetric(
+					"vault.token.count",
+					otel.ToggleScalarDataType,
+				),
+				otel.UpdateMetric(
+					"vault.core.request.count",
+					otel.ToggleScalarDataType,
+				),
+			),
+			otel.NormalizeSums(),
+			otel.MetricsTransform(
+				otel.AddPrefix("workload.googleapis.com"),
+			),
+		},
+	}}
+}
+
+type metricTransformer struct {
+	OldName     string
+	NewName     string
+	Description string
+	Unit        string
+}
+
+func (r MetricsReceiverVault) addStorageMetrics() (transforms []otel.TransformQuery, newMetrics []string) {
+	storageLabel := "storage"
+
+	storages := []string{
```

Comment:
Can we define all constants at the top of the file, makes it easier to update and less error prone.

Reply:
Updated

Reply:
[nit] Can we also move the list of storages to the top aswell and alphabetize it? Instead of reverse ahphabetical order? That is if the order doesn't matter.

Reply:
Updated